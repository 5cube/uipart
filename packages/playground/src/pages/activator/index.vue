<template>
  <h1>Activator</h1>
  <div class="py-6">
    <Modal>
      <template v-slot:activator>
        <Btn class="bg-transparent ring-1"> Activator </Btn>
      </template>
      <div class="bg-green-300 mt-4 p-8">Content</div>
    </Modal>
  </div>
  <div class="py-6">
    <Dialog v-model="dialog">
      <template v-slot:activator="{ attrs, listeners }">
        <Btn v-bind="attrs" v-on="listeners"> Activator dialog el </Btn>
      </template>
      <div class="bg-green-300 mt-4 p-8">
        Content
        <Btn @click="dialog = !dialog"> Close </Btn>
      </div>
    </Dialog>
  </div>
</template>

<script>
import { ref } from 'vue'
import Btn from '../../components/Btn'
import Modal from '../../components/Modal'
import Dialog from '../../components/Dialog.vue'
export default {
  components: {
    Btn,
    Modal,
    Dialog,
  },
  setup() {
    const dialog = ref(false)

    return {
      dialog,
    }
  },
}
</script>
