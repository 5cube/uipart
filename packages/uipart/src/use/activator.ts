import {
  ref,
  reactive,
  computed,
  cloneVNode,
  onMounted,
  onBeforeMount,
  onUnmounted,
  nextTick,
  watch,
} from 'vue'

import { SetupProps, SetupContext } from '../types'

import { useToggle } from './toggle'

import { parseEventName } from '../util/parse'

export const useActivatorProps = () => {
  return {
    modelValue: {
      type: [Boolean, String, Number],
      default: false,
    },
    activator: {
      type: [String, Object, HTMLElement, Element],
      default: null,
    },
    openOnHover: {
      type: Boolean,
      default: true,
    },
    openOnFocus: {
      type: Boolean,
      default: true,
    },
    openOnClick: {
      type: Boolean,
      default: true,
    },
    disableKeys: Boolean,
    disabled: Boolean,
  }
}

export interface ActivatorProps {
  modelValue: string | number | boolean | null | undefined
  activator: any
  disabled: boolean
  openOnHover: boolean
  openOnFocus: boolean
  openOnClick: boolean
  disableKeys: boolean
}
/**
 * @param {Object} props The props of use-case, readonly/reactive proxy.
 * @param {string|number|boolean} [props.modelValue]
 * @param {string|Object|HTMLElement|Element} [props.activator]
 * @param {boolean} [props.disabled]
 * @param {boolean} [props.openOnHover]
 * @param {boolean} [props.openOnFocus]
 * @param {boolean} [props.openOnClick]
 * @param {boolean} [props.disableKeys]
 */
export const useActivator = (props: ActivatorProps | SetupProps, { slots, emit }: SetupContext) => {
  const activatorNode = ref<any>(null)
  const activatorElement = ref<HTMLElement|null>(null)
  const _listeners = ref<any>({})

  const toggleProps = reactive({
    modelValue: computed(() => props.modelValue),
  })
  const { isActive } = useToggle(toggleProps, { emit })

  watch([() => props.activator, () => props.openOnHover], () => {
    resetActivator()
  })

  watch(_listeners, () => {
    nextTick(addActivatorEvents)
  })

  onMounted(() => {
    genActivatorListeners()
  })

  watch(() => props.disabled, (val) => {
    if (val) isActive.value = false
  })

  onBeforeMount(() => {
    nextTick(() => {
      if (props.modelValue) isActive.value = true
    })
  })

  onUnmounted(() => {
    removeActivatorEvents()
  })

  // Functions
  const resetActivator = () => {
    removeActivatorEvents()
    activatorElement.value = null
    getActivator()
    genActivatorListeners()
  }

  const genActivatorAttributes = ($attrs?: any) => {
    return Object.assign({
      role: 'button',
      'aria-haspopup': true,
      'aria-expanded': isActive.value ? true : undefined,
    }, $attrs)
  }

  const genActivatorListeners = ($listeners?: any) => {
    if (props.disabled) return {}

    const listeners:any = {}

    if (props.openOnHover) {
      listeners.onMouseenter = () => {
        isActive.value = true
      }
      listeners.onMouseleave = () => {
        isActive.value = false
      }
    } else {
      listeners.onClick = (e: MouseEvent) => {
        e.stopPropagation()

        isActive.value = !isActive.value
      }
    }

    if (props.openOnFocus) {
      listeners.onFocus = () => {
        isActive.value = true
      }
    }

    if (!props.disableKeys) {
      listeners.onKeydown = (e: KeyboardEvent) => {
        if (e.key === 'Esc' || e.key === 'Escape') {
          isActive.value = false
        }
      }
    }

    Object.assign(listeners, $listeners)

    _listeners.value = listeners

    return _listeners.value
  }

  const genActivator = ($attrs?: any, $listeners?: any) => {
    const attrs = genActivatorAttributes($attrs)
    const listeners = genActivatorListeners($listeners)
    const data = {
      ...attrs,
      ...listeners,
    }

    const listenersParsed = Object.keys(listeners).reduce((acc, key) => {
      const [type] = parseEventName(key)
      acc[type] = listeners[key]
      return acc
    }, {} as Record<string, any>)
    const node = slots.activator ? slots.activator({ attrs, listeners: listenersParsed }) : []

    // Auto merge data only in first node,
    // in other case set manually on v-slot attrs and listeners
    const [firstNode, ...others] = node
    const transformedNode = firstNode ? cloneVNode(firstNode, data) : undefined
    activatorNode.value = transformedNode
    return firstNode ? [transformedNode, ...others] : undefined
  }

  const addActivatorEvents = () => {
    const _activator: any = getActivator()
    if (
      !props.activator ||
      props.disabled ||
      !_activator
    ) return

    const keys = Object.keys(_listeners.value)
      .map(key => {
        const [type, modifiers] = parseEventName(key)
        return [key, type, modifiers]
      })

    for (const [key, type, modifiers] of keys) {
      _activator.addEventListener(type, _listeners.value[key], modifiers)
    }
  }

  const removeActivatorEvents = () => {
    if (
      !props.activator ||
      !activatorElement.value
    ) return

    const keys = Object.keys(_listeners.value)
      .map(key => {
        const [type, modifiers] = parseEventName(key)
        return [key, type, modifiers]
      })

    for (const [key, type, modifiers] of keys) {
      activatorElement.value.removeEventListener(type, _listeners.value[key], modifiers)
    }

    _listeners.value = {}
  }

  const getActivator = () => {
    if (activatorElement.value) return activatorElement.value

    let _activator = null

    if (props.activator) {
      const target = document

      if (typeof props.activator === 'string') {
        // Selector
        _activator = target.querySelector(props.activator)
      } else if (props.activator.$el) {
        // Component
        _activator = props.activator.$el
      } else {
        // HTMLElement | Element
        _activator = props.activator
      }
    } else if (activatorNode.value) {
      _activator = activatorNode.value.el
    }

    activatorElement.value = _activator

    return activatorElement.value
  }

  return {
    isActive,
    activatorElement,
    genActivatorAttributes,
    genActivatorListeners,
    genActivator,
    addActivatorEvents,
    removeActivatorEvents,
    getActivator,
  }
}
