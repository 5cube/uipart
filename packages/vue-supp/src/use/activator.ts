import {
  ref,
  reactive,
  computed,
  cloneVNode,
  onMounted,
  onBeforeMount,
  onUnmounted,
  nextTick,
  watch,
  PropType,
  VNode,
} from 'vue'

import { SetupContext } from '../../types'

import { toggle } from './toggle'

import parseEventName from '../utils/parseEventName'

export interface ActivatorProps {
  modelValue: string | number | boolean | null | undefined
  activator?: any
  disabled?: boolean
  openOnHover: boolean
  openOnFocus: boolean
  openOnClick: boolean
  disableKeys?: boolean
}

export const useActivatorProps = () => {
  return {
    modelValue: {
      type: [Boolean, String, Number] as PropType<
        string | number | boolean | null | undefined
      >,
      default: false,
    },
    activator: {
      default: (null as unknown) as PropType<
        string | HTMLElement | VNode | Element | null
      >,
      validator: (val: string | Record<string, unknown>) => {
        return ['string', 'object'].includes(typeof val)
      },
    },
    openOnHover: {
      type: Boolean,
      default: true,
    },
    openOnFocus: {
      type: Boolean,
      default: true,
    },
    openOnClick: {
      type: Boolean,
      default: true,
    },
    disableKeys: Boolean,
    disabled: Boolean,
  }
}

export const useActivator = (
  props: ActivatorProps,
  { slots, emit }: Pick<SetupContext, 'slots' | 'emit'>
) => {
  const activatorNode = ref<any>()
  const activatorElement = ref<HTMLElement>()
  const _listeners = ref<any>({})

  const { useToggle } = toggle()

  const toggleProps = reactive({
    modelValue: computed(() => props.modelValue),
  })
  const { isActive } = useToggle(toggleProps, { emit })

  watch([() => props.activator, () => props.openOnHover], () => {
    resetActivator()
  })

  watch(_listeners, () => {
    nextTick(addActivatorEvents)
  })

  onMounted(() => {
    genActivatorListeners()
  })

  watch(
    () => props.disabled,
    (val) => {
      if (val) isActive.value = false
    }
  )

  onBeforeMount(() => {
    nextTick(() => {
      if (props.modelValue) isActive.value = true
    })
  })

  onUnmounted(() => {
    removeActivatorEvents()
  })

  // Functions
  const resetActivator = () => {
    removeActivatorEvents()
    activatorElement.value = undefined
    getActivator()
    genActivatorListeners()
  }

  const genActivatorAttributes = ($attrs?: any) => {
    return Object.assign(
      {
        role: 'button',
        'aria-haspopup': true,
        'aria-expanded': isActive.value ? true : undefined,
      },
      $attrs
    )
  }

  const genActivatorListeners = ($listeners?: any) => {
    if (props.disabled) return {}

    const listeners: any = {}

    if (props.openOnHover) {
      listeners.onMouseenter = () => {
        isActive.value = true
      }
      listeners.onMouseleave = () => {
        isActive.value = false
      }
    } else {
      listeners.onClick = (e: MouseEvent) => {
        e.stopPropagation()

        isActive.value = !isActive.value
      }
    }

    if (props.openOnFocus) {
      listeners.onFocus = () => {
        isActive.value = true
      }
    }

    if (!props.disableKeys) {
      listeners.onKeydown = (e: KeyboardEvent) => {
        if (e.key === 'Esc' || e.key === 'Escape') {
          isActive.value = false
        }
      }
    }

    Object.assign(listeners, $listeners)

    _listeners.value = listeners

    return _listeners.value
  }

  const genActivator = ($attrs?: any, $listeners?: any) => {
    const attrs = genActivatorAttributes($attrs)
    const listeners = genActivatorListeners($listeners)
    const data = {
      ...attrs,
      ...listeners,
    }

    const listenersParsed = Object.keys(listeners).reduce((acc, key) => {
      const [type] = parseEventName(key)
      acc[type] = listeners[key]
      return acc
    }, {} as Record<string, any>)
    const node = slots.activator
      ? slots.activator({ attrs, listeners: listenersParsed })
      : []

    // Auto merge data only in first node,
    // in other case set manually on v-slot attrs and listeners
    const [firstNode, ...others] = node
    const transformedNode = firstNode ? cloneVNode(firstNode, data) : undefined
    activatorNode.value = transformedNode
    return firstNode ? [transformedNode, ...others] : undefined
  }

  const addActivatorEvents = () => {
    const _activator: any = getActivator()
    if (!props.activator || props.disabled || !_activator) return

    const keys = Object.keys(_listeners.value).map((key) => {
      const [type, modifiers] = parseEventName(key)
      return [key, type, modifiers]
    })

    for (const [key, type, modifiers] of keys) {
      _activator.addEventListener(
        type as any,
        _listeners.value[key as string],
        modifiers
      )
    }
  }

  const removeActivatorEvents = () => {
    if (!props.activator || !activatorElement.value) return

    const keys = Object.keys(_listeners.value).map((key) => {
      const [type, modifiers] = parseEventName(key)
      return [key, type, modifiers]
    })

    for (const [key, type, modifiers] of keys) {
      activatorElement.value.removeEventListener(
        type as any,
        _listeners.value[key as string],
        modifiers as any
      )
    }

    _listeners.value = {}
  }

  const getActivator = () => {
    if (activatorElement.value) return activatorElement.value

    let _activator = null

    if (props.activator) {
      const target = document

      if (typeof props.activator === 'string') {
        // Selector
        _activator = target.querySelector(props.activator)
      } else if (props.activator.$el) {
        // Component
        _activator = props.activator.$el
      } else {
        // HTMLElement | Element
        _activator = props.activator
      }
    } else if (
      activatorNode.value &&
      activatorNode.value.key === '_activator' &&
      activatorNode.value.children.length
    ) {
      // set correct element on nested activator slot
      // TODO: select element recursively if element type is 'Symbol' or 'text'
      _activator = activatorNode.value.children[0].el
    } else if (activatorNode.value) {
      _activator = activatorNode.value.el
    }

    activatorElement.value = _activator

    return activatorElement.value
  }

  return {
    isActive,
    activatorElement,
    genActivatorAttributes,
    genActivatorListeners,
    genActivator,
    addActivatorEvents,
    removeActivatorEvents,
    getActivator,
  }
}
