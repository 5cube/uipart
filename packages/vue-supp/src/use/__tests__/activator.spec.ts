import { mount } from '@vue/test-utils'
import { h, defineComponent } from 'vue'
import { useActivatorProps, useActivator } from '../activator'

describe('activator.ts', () => {
  const component = defineComponent({
    props: {
      ...useActivatorProps(),
      attrs: {
        type: Object,
        default: () => ({}),
      },
      listeners: {
        type: Object,
        default: () => ({}),
      },
    },

    setup(props, ctx) {
      const instance = useActivator(props, ctx)

      if (props.attrs) instance.setActivatorAttrs(props.attrs)
      if (props.listeners) instance.setActivatorListeners(props.listeners)

      return {
        ...instance,
      }
    },

    render() {
      return this.genActivator()
    },
  })

  function mountFunction(options = {}) {
    return mount(component, {
      ...options,
    })
  }

  beforeEach(() => {
    document.body.innerHTML = '<div id="app"></div>'
  })

  it('should render activator with slot and match snapshot', () => {
    const wrapper = mountFunction({
      slots: {
        activator: () => h('button', 'Activator'),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correct attrs/listeners and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        attrs: {
          custom: 'custom-attr',
        },
        listeners: {
          onKeypress: () => undefined,
        },
      },
      slots: {
        activator: () => h('button', 'Activator'),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.vm.genActivatorListeners()).toHaveProperty('onKeypress')
  })

  it('should add custom attrs/listeners from gen function', () => {
    const wrapper = mountFunction({
      props: {
        attrs: {
          custom: 'custom-attr',
        },
        listeners: {
          onKeypress: () => undefined,
        },
      },
      slots: {
        activator: () => h('button', 'Activator'),
      },
    })

    const attrs = wrapper.vm.genActivatorAttributes({
      more: 'more-attr',
    })
    const listeners = wrapper.vm.genActivatorListeners({
      onKeyup: () => undefined,
    })
    expect(attrs).toHaveProperty('more')
    expect(listeners).toHaveProperty('onKeyup')
  })

  it('should pass correct data to v-slot, { attrs, listeners }', () => {
    const wrapper = mountFunction({
      slots: {
        activator: ({ attrs, listeners }: any) =>
          h(
            'button',
            [JSON.stringify(attrs), Object.keys(listeners)].join(',')
          ),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it.todo(
    'should auto merge attrs and listeners when activator slot is not nested and match snapshot'
  )

  it('should set correct activator element from prop', async () => {
    const wrapperComponent = defineComponent({
      props: useActivatorProps(),
      setup(props) {
        return () =>
          h('div', [
            h('button', { id: 'activator' }, 'Activator'),
            h(component, props),
          ])
      },
    })

    const wrapper = mount(wrapperComponent, {
      attachTo: '#app',
      props: { activator: '#activator' },
    })

    const comp = wrapper.findComponent(component)
    await wrapper.get('#activator').trigger('focus')
    expect(comp.vm.isActive).toBe(true)
  })

  it('should reset activator prop change', async () => {
    const wrapperComponent = defineComponent({
      props: useActivatorProps(),
      setup(props) {
        return () =>
          h('div', [
            h('button', { id: 'activator-1' }, 'Activator 1'),
            h('button', { id: 'activator-2' }, 'Activator 2'),
            h(component, props),
          ])
      },
    })

    const wrapper = mount(wrapperComponent, {
      attachTo: '#app',
      props: { activator: '#activator-1' },
    })

    const comp = wrapper.findComponent(component)
    await wrapper.setProps({ activator: '#activator-2' })
    await wrapper.get('#activator-2').trigger('focus')
    expect(comp.vm.isActive).toBe(true)
  })

  it('should reset listeners on openOnHover, openOnFocus, openOnClick, disableKeys prop change', async () => {
    const wrapperComponent = defineComponent({
      props: useActivatorProps(),
      setup(props) {
        return () =>
          h('div', [
            h('button', { id: 'activator' }, 'Activator'),
            h(component, props),
          ])
      },
    })

    const wrapper = mount(wrapperComponent, {
      attachTo: '#app',
      props: { activator: '#activator' },
    })

    const comp = wrapper.findComponent(component)
    const activator = wrapper.get('#activator')
    await wrapper.setProps({ openOnHover: false })
    await activator.trigger('mouseenter')
    expect(comp.vm.isActive).toBe(false)

    await activator.trigger('click')
    expect(comp.vm.isActive).toBe(true)
    await activator.trigger('click')
    expect(comp.vm.isActive).toBe(false)

    await wrapper.setProps({ openOnClick: false })
    await activator.trigger('click')
    expect(comp.vm.isActive).toBe(false)

    await wrapper.setProps({ openOnFocus: false })
    await activator.trigger('focus')
    expect(comp.vm.isActive).toBe(false)

    comp.vm.isActive = true
    await activator.trigger('keydown', { key: 'Esc' })
    expect(comp.vm.isActive).toBe(false)
    await wrapper.setProps({ disableKeys: true })
    comp.vm.isActive = true
    await activator.trigger('keydown', { key: 'Esc' })
    expect(comp.vm.isActive).toBe(true)
  })

  it('should immediate render on modelValue prop is active', () => {
    const wrapper = mountFunction({
      props: {
        modelValue: true,
      },
    })

    expect(wrapper.vm.isActive).toBe(true)
  })

  it('should deactivate on disable prop change', async () => {
    const wrapper = mountFunction({
      props: {
        modelValue: true,
      },
    })

    expect(wrapper.vm.isActive).toBe(true)
    await wrapper.setProps({ modelValue: true, disabled: true })
    expect(wrapper.vm.isActive).toBe(false)
  })

  it('should toggle isActive and emit update:modelValue', async () => {
    const wrapper = mountFunction()

    await wrapper.setProps({ modelValue: true })
    expect(wrapper.emitted()['update:modelValue'].length).toBe(1)
  })

  it('should focus first element with role="button" or aria-haspopup="true"', () => {
    const wrapper = mountFunction({
      slots: {
        activator: () => [
          h('button', { id: 'btn-1' }, 'Button 1'),
          h('button', { id: 'btn-2' }, 'Button 2'),
        ],
      },
      attachTo: '#app',
    })

    const btn1 = wrapper.get('#btn-1')
    expect(document.activeElement).not.toEqual(btn1.element)
    wrapper.vm.focusActivator()
    expect(document.activeElement).toEqual(btn1.element)
  })
})
