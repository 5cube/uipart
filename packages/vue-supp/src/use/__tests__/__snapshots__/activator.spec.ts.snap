// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activator.ts should pass correct data to v-slot, { attrs, listeners } 1`] = `<button role="button" aria-haspopup="true">{"role":"button","aria-haspopup":true},mouseenter,mouseleave,focus,keydown</button>`;

exports[`activator.ts should render activator with slot and match snapshot 1`] = `<button role="button" aria-haspopup="true">Activator</button>`;

exports[`activator.ts should render correct attrs/listeners and match snapshot 1`] = `<button role="button" aria-haspopup="true" custom="custom-attr">Activator</button>`;
